## Topic: Recursion

### Write recursive python program for the following:
### 1. octal to binary

In [27]:
num = int(input("Given Octal Number: "))

def octToDec(num, pow):
    rem = num % 10
    num = num // 10
    digit = rem * (8 ** pow)
    pow += 1
    
    # end condition
    if num == 0:
        return digit
    
    # else keep adding
    return digit + octToDec(num, pow)

def decToBin(num):
    
    # end condition
    if num == 0:
        return
    
    # recursion
    decToBin(num//2)
    print(num%2, end = '')
    

def octToBin(num):
    dec = octToDec(num, 0)
    print("Equivalent Binary Conversion: ", end = '')
    decToBin(dec)

octToBin(num)

Given Octal Number: 171
Equivalent Binary Conversion: 1111001

### 2. hexadecimal to binary

In [26]:
numstr = input("Given Hexadecimal Number: ")


def hexToDec(numStr, pow):
    
    # end condition
    if numStr == '':
        return 0
    
    # extract each digit starting from the end
    hexDigitStr = numStr[-1]
    numstr = numStr[:-1]
    hexDigit = 0
    
    # convert Alpha to Digits using ASCII values
    if hexDigitStr.isalpha():
        hexDigit = ord(hexDigitStr) - 55
    else:
        hexDigit = int(hexDigitStr)
    
    decDigit = hexDigit * (16 ** pow)
    pow += 1
    
    # else keep adding
    return decDigit + hexToDec(numstr, pow)


def decToBin(num):
    
    # end condition
    if num == 0:
        return
    
    decToBin(num//2)
    print(num%2, end = '')
    

def hexToBin(num):
    dec = hexToDec(num, 0)
    print(f"Equivalent Decimal Number: {dec}")
    print("Equivalent Binary Number: ", end = '')
    decToBin(dec)

    
hexToBin(numstr)

Given Hexadecimal Number: AB123
Equivalent Decimal Number: 700707
Equivalent Binary Number: 10101011000100100011

### 3. binary to octal

#### Working:
input binary number as a string, divide them into parts of 3 - 3 digits, and then convert each to octal individually<br>
Example Input: 1111001<br>
Divide them into groups of 3 <ins>starting from the end</ins>: 001   111   001<br>
Convert each to octal, individually: 171<br>

In [36]:
binNum = int(input("Given Binary Number: "))

# convert 3 digit binary number to decimal number
def binToDec(binDigit):
    digit3 = (binDigit // 100) * (2 ** 2)
    binDigit = binDigit % 100
    
    digit2 = (binDigit // 10) * (2 ** 1)
    binDigit = binDigit % 10
    
    digit1 = binDigit * (2 ** 0)
    
    return digit3 + digit2 + digit1

def binToOct(binNum):
    
    # end condition
    if binNum == 0:
        return
    
    # extract last 3 digits from binary
    binDigit = binNum % 1000
    binNum = binNum // 1000
    
    # convert digit to decimal
    decDigit = binToDec(binDigit)
    
    # recursion
    binToOct(binNum)
    print(decDigit, end = '')
    
print("Equivalent Octal Number: ", end = '')
binToOct(binNum)

Given Binary Number: 111000111
Equivalent Octal Number: 707

### 4. Enter a 4-digit number and display them in words

In [4]:
def digitToWord(num):
    # base condition
    if num == 0:
        return
    
    # extract each digit
    digit = num % 10
    num = num // 10
    
    # recursive call
    digitToWord(num)
    print(numToWord[digit], end = ' ')
    

num = int(input("Given 4-digit number: "))
numToWord = {1: "one", 2: "two", 3: "three", 4: "four", 5: "five", 6: "six", 7: "seven", 8: "eight", 9: "nine", 0: "zero"} 
print("Given number in words: ", end = '')
digitToWord(num)

Given 4-digit number: 4301
Given number in words: four three zero one 

### 5. Find the sum of the series using Recursion

### (a)  x - x^3/3! + x^5/5! - x^7/7! + ... 

In [16]:
x = int(input("Value of x: "))
terms = int(input("Number of terms: "))

def fact(num):
    if num == 0 or num == 1:
        return 1
    return num * fact(num - 1)

def sum1(terms, x, pow, sign):
    # base condition
    if terms == 0:
        return 0
    
    return sign * ((x ** pow) / fact(pow)) + sum1(terms - 1, x, pow + 2, sign * -1)

print(f"sin {x} using expansion till {terms} terms: {sum1(terms, x, 1, 1)}")

Value of x: 90
Number of terms: 10
sin 90 using expansion till 10 terms: -1.0651331796981349e+20


### (b)  1 - x^2/2! + x^4/4! - x^6/6! + ... 

In [17]:
x = int(input("Value of x: "))
terms = int(input("Number of terms: "))

def fact(num):
    if num == 0 or num == 1:
        return 1
    return num * fact(num - 1)

def sum2(terms, x, pow, sign):
    # base condition
    if terms == 0:
        return 0
    
    return sign * ((x ** pow) / fact(pow)) + sum2(terms - 1, x, pow + 2, sign * -1)

print(f"cos {x} using expansion till {terms} terms: {sum2(terms, x, 0, 1)}")

Value of x: 0
Number of terms: 10
cos 0 using expansion till 10 terms: 1.0


### (c)  1 + x + x^2/2! + x^3/3!  ... 

In [18]:
x = int(input("Value of x: "))
terms = int(input("Number of terms: "))

def fact(num):
    if num == 0 or num == 1:
        return 1
    return num * fact(num - 1)

def sum3(terms, x, pow):
    # base condition
    if terms == 0:
        return 0
    
    return ((x ** pow) / fact(pow)) + sum3(terms - 1, x, pow + 1)

print(f"Sum of series using expansion till {terms} terms: {sum3(terms, x, 0)}")

Value of x: 2
Number of terms: 3
Sum of series using expansion till 3 terms: 5.0


### (d)  x - x^2/2 + x^3/3 - x^4/4 + x^5/5  ... 

In [19]:
x = int(input("Value of x: "))
terms = int(input("Number of terms: "))

def sum4(terms, x, pow, sign):
    # base condition
    if terms == 0:
        return 0
    
    return sign * ((x ** pow) / pow) + sum4(terms - 1, x, pow + 2, sign * -1)

print(f"log(1 + {x}) using expansion till {terms} terms: {sum4(terms, x, 1, 1)}")

Value of x: 2
Number of terms: 10
log(1 + 2) using expansion till 10 terms: -21580.212413523863


### (e)  x - x^2/2 - x^3/3 - x^4/4 - x^5/5  ... 

In [20]:
x = int(input("Value of x: "))
terms = int(input("Number of terms: "))

def sum5(terms, x, pow):
    # base condition
    if terms == 0:
        return 0
    
    return ((x ** pow) / pow) - sum5(terms - 1, x, pow + 2)

print(f"log(1 + {x}) using expansion till {terms} terms: {sum5(terms, x, 1)}")

Value of x: 2
Number of terms: 10
log(1 + 2) using expansion till 10 terms: -21580.212413523863


### 6. Define Ackermann's function: A(m,n) = 
###                                                        (n + 1), if m = 0
###                                                         A(m - 1, 1) if m > 0 & n = 0
###                                                         A(m - 1, A(m, n - 1)), if m > 0 & n > 0

In [21]:
m = int(input("m: "))
n = int(input("n: "))

def A(m, n):
    # base condition
    if m == 0:
        return (n + 1)
    
    elif m > 0 and n == 0:
        return A(m - 1, 1)
    
    elif m > 0 and n > 0:
        return A(m - 1, A(m, n - 1))
    
A(m, n)

m: 3
n: 3


61